---
title: "Positive Control Experiment"
author: "Alyssa Frazee"
date: "August 15, 2014"
output: html_document
---

The previous experiment, the negative control, showed that Ballgown's default statistical tests are appropriately conservative when there is no signal present in the data. Here we do an experiment to show that these statistical tests are capable of making discoveries when differential expression really is present.

For this experiment you will need `fpkm.rda` ([download here](http://files.figshare.com/1625419/fpkm.rda)). 

```{r, message=FALSE, warning=FALSE}
library(ballgown)
load('fpkm.rda')
head(pData(fpkm))
```

Sex information was stored externally; add to object:

```{r addsex}
pheno_table = read.table('../GEUVADIS_preprocessing/pop_data_withuniqueid.txt', sep='\t', header=TRUE)
pd = pData(fpkm)
pd$sex = pheno_table$sex[match(pd$IndividualID, pheno_table$hapmap_id)]
pData(fpkm) = pd
```

For the positive control, we'll analyze differential expression between males and females on the Y chromosome, using individuals from the FIN population (same as negative control experiment). 

```{r, warning=FALSE, message=FALSE}
fin = subset(fpkm, 'population == "FIN" & UseThisDup == 1', genomesubset=FALSE)
finy = subset(fin, 'chr == "chrY"', genomesubset=TRUE)
finy
table(pData(finy)$sex)
```

```{r, echo=FALSE}
# write out table for use with Cuffdiff
write.table(pData(finy)[,c('dirname', 'sex')], file='sex_info.txt', col.names=FALSE, row.names=FALSE, quote=FALSE)
```

So we are doing a 58 vs. 37 comparison, with 58 females in the analysis, on 433 assembled Y-chromosome transcripts:

```{r}
positive_results_all = stattest(finy, feature='transcript', meas='FPKM', covariate='sex')
```

We will only keep/analyze results from transcripts whose mean FPKM value across the male FIN samples is greater than 0.01 (and re-adjust p-values accordingly):

```{r}
male_means = rowMeans(texpr(finy)[,pData(finy)$sex=='male'])
positive_results = positive_results_all[male_means >= 0.01,]
positive_results$qval = p.adjust(positive_results$pval, 'fdr')
summary(positive_results$qval)
sum(positive_results$qval < 0.05) / nrow(positive_results)
sum(positive_results$qval < 0.2) / nrow(positive_results)
hist(positive_results$pval, breaks=30, col='gray', xlab='p-values', main='Y-chromosome p-values')
write.csv(positive_results$pval, file='1b.csv')
```

```{r}
pdf('figure1b.pdf')
hist(positive_results$pval, breaks=30, col='gray', xlab='p-values', main='Y-chromosome p-values')
dev.off()
```

Here the p-value histogram shows a very strong signal, as was to be expected, and the proportion of q-values below 0.2 is `r sum(positive_results$qval < 0.2) / nrow(positive_results)`, showing that Ballgown's statistical significance estimates render researchers capable of detecting real differential expression.

Cuffdiff analysis is reported in a separate file.

We also compare to the results of an EBSeq analysis:

```{r, cache=TRUE}
library(EBSeq)
library(reshape2)
rdata = read.table('Y_isoforms.read_group_tracking.gz', header=TRUE, sep='\t')
Data = acast(rdata, formula=tracking_id~replicate+condition, value.var='raw_frags')
Conditions = ballgown:::ss(colnames(Data), pattern='_', slot=2)
IsoformNames = rownames(Data)
iso_gene_relationship = read.table('isoform_exp.diff', 
    colClasses=c('character', 'character', rep('NULL', 12)), header=TRUE)
sum(IsoformNames != iso_gene_relationship$test_id) # expect 0
IsosGeneNames = iso_gene_relationship$gene_id
IsoSizes = MedianNorm(Data)
NgList = GetNg(IsoformNames, IsosGeneNames)
IsoNgTrun = NgList$IsoformNgTrun

system.time(IsoEBOut <- EBTest(Data=Data, NgVector=IsoNgTrun, 
    Conditions=as.factor(Conditions), sizeFactors=IsoSizes, maxround=20))

# check convergence
IsoEBOut$Alpha
IsoEBOut$Beta
IsoEBOut$P
```

We see that EBSeq called `r sum(IsoEBOut$PPDE > 0.95)` transcripts (`r 100*sum(IsoEBOut$PPDE > 0.95) / length(IsoEBOut$PPDE)` percent) differentially expressed (posterior probability of being differentially expressed at least 0.95), and if we move the PPDE cutoff to 0.8, EBSeq called `r sum(IsoEBOut$PPDE > 0.8)` transcripts (`r 100*sum(IsoEBOut$PPDE > 0.8) / length(IsoEBOut$PPDE)` percent) differerentially expressed.


```{r, results='markup'}
sessionInfo()
```


